src/passes/Memory64Lowering.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2020 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Lowers a module with loads and stores that access a 64-bit memory with
+// one that works as-is on wasm32.
+//
+
+#include "ir/bits.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+struct Memory64Lowering : public WalkerPass<PostWalker<Memory64Lowering>> {
+
+  void WrapIfAddress64(Expression*& ptr) {
+    auto& module = *getModule();
+    if (!module.memory.is64()) {
+      return;
+    }
```

Comment:
It would be good to do this in a `doWalkFunction` implementation to avoid the traversal entirely when possible.

```c++
void doWalkFunction(Function* func) {
  if (getModule()->memory.is64()) {
    super::doWalkFunction(func);
  }
}
```

Reply:
Or skip the entire module?

Reply:
This is not a pass that is by default on, it will be something people select because they specifically want this lowering. But either way works for me.

How do I skip the entire module? `doWalkFunction(Module*)` I guess?

Reply:
You'll want to override `void run(PassRunner* runner, Module* module)`. If wasm64 is enabled, you can have it call `super::run(runner, module)`.

Reply:
I guess this is all going to blow up when we support accessing multiple memories at once, but ok..